src/fs/service.go

```diff
@@ -626,17 +629,41 @@ func (service *MercuryFsService) uploadFile(writer http.ResponseWriter, request
 		}
 		defer file.Close()
 
-		// FIXME -- check the filename so it does not start with dots, or slashes!
+		//check if the file name is valid
+		if !validFilename(handler.Filename) {
+			debug(2, "Error prefix of uploaded file")
+			writer.WriteHeader(http.StatusUnsupportedMediaType)
+			service.debugInfo.requestServed(int64(0))
+			log("\"POST %s\" 415 0 \"%s\"", query, ua)
+			return
+		}
+
 		fullPath, _ := service.fullPathToFile(share, path+"/"+handler.Filename)
 
-		f, err := os.OpenFile(fullPath, os.O_WRONLY|os.O_CREATE, 0644)
+		md5 := CalMD5(file)
```

Comment:
This should not be calculated here, calculate MD5 for the uploaded file only if another file exists with the same name.